### PR TITLE
Make hyperlinks work

### PR DESCRIPTION
**Pull request by BarryNolte**
Links now work in D2 preview.  Links will not work in D2 diagrams embedded in markdown.  I'm trying to find a 'correct' way to do this.  FWIW - links in markdown that's embedded in D2 also don't work.

### CODE REVIEW

File: pages/previewPage.html [hunk 1]
```diff
@@ -252,6 +253,22 @@
...
+        (event) => {
+          var domNode = event.target;
+          while (domNode !== null) {
+            console.log(domNode.tagName);
+            if (domNode.tagName.toLowerCase() === "a" && domNode.href !== null) {
+              vscode.postMessage({ command: "clickOnTag_A", link: domNode.href.baseVal });
+              return;
+            }
+            domNode = domNode.parentNode;
+          }
```
Comment (gavin-ts) [inline pages/previewPage.html:268]: I think you can use this for simplicity:

`const link = event.target.closest("a[href]")`

Comment (BarryNolte) [reply pages/previewPage.html]: I learn something new every day.

File: src/browserWindow.ts [hunk 2]
```diff
@@ -46,6 +48,39 @@ export class BrowserWindow {
...
+
+    this.webViewPanel.webview.onDidReceiveMessage(
+      (message) => {
+        switch (message.command) {
+          case "clickOnTag_A": {
+            const uri = Uri.parse(message.link);
+            switch (uri.scheme) {
+              case "file":
+                {
+                  const filepath = path.join(filePath, message.link);
```
Comment (gavin-ts) [inline src/browserWindow.ts:60]: can you check to see if it is an absolute path first, and only join for relative paths?

**Review by gavin-ts — CHANGES_REQUESTED** [review]
tested and this works, but I'm thinking it might be better to have it open in the default browser/application rather than opening in vscode if possible

File: src/browserWindow.ts [hunk 3]
```diff
@@ -46,6 +48,39 @@ export class BrowserWindow {
...
+                      window.showTextDocument(document);
+                    },
+                    () => {
+                      window.showErrorMessage(`Could not open: ${filepath}`);
+                    }
+                  );
+                }
+                break;
+              // Do nothing, vscode opens hyperlink on it's own
+              case "html":
```
[the remark this answers is not in the document]
Comment (BarryNolte) [reply src/browserWindow.ts]: Greatly simplified file vs http detection

File: src/browserWindow.ts (hunk 2 above)
Comment (BarryNolte) [reply src/browserWindow.ts]: Took care of this.

**Review by gavin-ts — COMMENTED** [review]
this works for text type files, (`.md, .d2, .svg` seem to work) but not `.png`. I think it is related to using `openTextDocument` specifically. Can we update this to use https://github.com/sindresorhus/open#opentarget-options to open the file in default application? Maybe there can be a user config setting if they want to try opening text files in vscode first vs default of always open in default application?

**Comment by BarryNolte** [comment]
I kept the logic the same, except when vscode can't open the file, then I'll pass it to shell.open,() or shell.default() as it's defined, then the default app will (hopefully) be invoked for that file type.  This will also launch executable files.  This can't be detected, 'for sure', on windows since executables don't have that attribute associated with them.  If that is an issue, then some heuristics can be inserted to hit the 99% case to at least say, 'this is an executable, are you sure you want to launch'.

File: src/browserWindow.ts [hunk 4]
```diff
@@ -76,14 +76,11 @@ export class BrowserWindow {
...
-                    // do nothing
-                  },
-                  () => {
-                    window.showErrorMessage(`Could not open: ${filepath}`);
-                  }
-                );
+                if (!existsSync(filepath)) {
+                  window.showErrorMessage(`File does not exist: ${filepath}`);
+                } else {
+                  shell.default(filepath);
```
Comment (gavin-ts) [inline src/browserWindow.ts:82]: why is this importing `*` as `shell` and `default`? why not just `import open from "open"` and `open.open(filepath)`?

**Comment by gavin-ts** [comment]
> I kept the logic the same, except when vscode can't open the file, then I'll pass it to shell.open,() or shell.default() as it's defined, then the default app will (hopefully) be invoked for that file type. This will also launch executable files. This can't be detected, 'for sure', on windows since executables don't have that attribute associated with them. If that is an issue, then some heuristics can be inserted to hit the 99% case to at least say, 'this is an executable, are you sure you want to launch'.

Hmm will people always want to have text files open in vscode? for examples svgs work with openTextDocument, but I'd prefer to have it open in my browser. I think it should be a configured setting if we add that feature

**Comment by BarryNolte** [comment]
I suppose I could add a checkbox of, "open in vscode if you can", otherwise let the OS sort it out.

File: src/browserWindow.ts (hunk 4 above)
Comment (BarryNolte) [reply src/browserWindow.ts:82]: This is what I got when I imported the module you referenced.

Comment (gavin-ts) [reply src/browserWindow.ts:82]: from https://github.com/sindresorhus/open#usage, I think you can just: 
```js
import open from 'open';

// Opens the URL in the default browser.
open(filepath);
```

Comment (BarryNolte) [reply src/browserWindow.ts]: That worked.  Six ways to import, none of them feel right...

**Review by gavin-ts — APPROVED** [review]
nice this works well, thinking about it opening in vscode if possible makes sense (user can use the vscode reveal in finder command if they want to open another way); if in the future anyone wants to turn it off in a setting we could add that then

**Pull request merged**